core/src/main/java/com/linecorp/armeria/server/RoutingPredicate.java

```diff
@@ -108,37 +114,124 @@
     @VisibleForTesting
     static <T, U> RoutingPredicate<T> of(String predicateExpr,
                                          Function<String, U> nameConverter,
-                                         Function<U, Predicate<T>> containsPredicateFactory,
-                                         BiFunction<U, String, Predicate<T>> equalsPredicateFactory) {
+                                         Function<U, Predicate<T>> containsPredicate,
+                                         BiFunction<U, String, Predicate<T>> equalsPredicate) {
+        final Optional<NamedPredicate<T>> namedPredicate;
+        final Optional<NamedPredicate<T>> orNamedPredicate = buildOrNamedPredicate(predicateExpr,
+                                                                                   nameConverter,
+                                                                                   containsPredicate,
+                                                                                   equalsPredicate);
+        if (orNamedPredicate.isPresent()) {
+            namedPredicate = orNamedPredicate;
+        } else {
+            namedPredicate = buildSingleExprNamedPredicate(predicateExpr,
+                                                           nameConverter,
+                                                           containsPredicate,
+                                                           equalsPredicate);
+        }
+        checkArgument(namedPredicate.isPresent(),
+                      "Invalid predicate: %s (expected: '%s', '%s' or '%s')",
+                      predicateExpr, CONTAIN_PATTERN.pattern(), COMPARE_PATTERN.pattern(),
+                      OR_PATTERN.pattern());
+        return new RoutingPredicate<>(namedPredicate.get().name, namedPredicate.get().predicate);
+    }
+
+    private static <T, U> Optional<NamedPredicate<T>> buildOrNamedPredicate(String predicateExpr,
+                                                                            Function<String, U> nameConverter,
+                                                                            Function<U, Predicate<T>>
+                                                                                    containsPredicate,
+                                                                            BiFunction<U, String, Predicate<T>>
+                                                                                    equalsPredicate) {
+        final Matcher orMatcher = OR_PATTERN.matcher(predicateExpr);
+        if (orMatcher.find()) {
+            checkArgument(!predicateExpr.endsWith("|"), "Invalid predicate: %s (expected: '%s')",
+                          predicateExpr, TRAILING_PIPE_PATTERN.matcher(predicateExpr).replaceAll(""));
```

Comment:
Should we use `if` statement to avoid the extra computation of `TRAILING_PIPE_PATTERN`?